### DISCUSSION TITLE
Add custom unit system support

### DISCUSSION THREAD

**Discussion by avee87** [home-assistant/architecture#682] [General]
Currently Home Assistant provides only 2 unit systems - Metric and Imperial with fixed measurements for each type. However, not all countries use exactly the same sets of units - you may want to use Celcius and miles at the same time, or mmHg for pressure, or something else. It becomes particularly important now with more measurements being added to the units system (thanks @rianadon) which get auto-converted to units potentially undesirable by user.

Would be great if users had granular control over units for each particular measurement. Essentially, add "Custom" measurement system where user would choose each measurement from a list of options. They would default to the previous system so it won't be hard to just override one unit rather than esigning the whole system from scratch.

With the current way unit systems are implemented it shouldn't be particularly hard to add a dynamic one where user can configure each measurement - essentially, add a dictionary argument [here](https://github.com/home-assistant/core/blob/dev/homeassistant/core.py#L1670-L1697) and use it to contruct a new system.

<img width="578" alt="Screenshot 2021-12-11 at 19 07 36" src="https://user-images.githubusercontent.com/6134677/145688642-6dab4050-d3bb-4595-bbe0-7a2f319d07c7.png">

**Comment by rianadon** [comment]
It would be great to have something like this!

Another idea I had would be to simply add more options. Below is what the darksky api used to use (their SI=our metric, us=our imperial, ca and uk2 are the two additional options):

<img width="702" alt="image" src="https://user-images.githubusercontent.com/18223213/145691064-080aefa3-e9a4-479d-969c-88219ac71b88.png">


I think the custom units approach works great for YAML, but configuring custom units in the UI would take up a lot of space on the settings page.

**Reply by avee87** [reply]
UI could be tucked away in a dialog or accordion so you don't see it unless you're actually updating it which probably doesn't happen very often.

**Reply by rianadon** [reply]
I think this would be a good way to present them. However, the frontend code becomes more complex at this point. So it's a tradeoff between keep frontend simple / allow granular configuration.

**Reply by farmio** [reply]
Another valuable unit selection would be for color temperature. HA uses `mired` versus a lot of bulbs and integration use `kelvin` natively.

**Comment by avee87** [comment]
PR with proposed changes to HA Core: https://github.com/home-assistant/core/pull/63344

**Comment by frenck** [comment]
I actually don't agree with this proposal. I rather see the possibility to be able to change the unit on devices/entities.

Creating an endless configuration for all types of data points and preferred measurements in the core configuration, is in my opinion, an endless game.

**Reply by avee87** [reply]
Isn't centralising units the point of having a unit system? I.e., instead of manually setting C or F on all your temperature sensors you set it once in config and have it recalculated properly. And if we do have such a unit system with recalculations, not being able to customise it is a significant limitation.

I don't know how many unit conversions will HA support in the future but at the moment it doesn't seem endless at all.

**Reply by frenck** [reply]
The list can become pretty big. Temperatures, pressure, speed, voltage, energy, power... and that is not just the main issue; one quickly goes into subcategories for those. For example, the speed of a car vs wind speed. Battery voltage, vs the voltage of the grid. (+ whatever is not covered by our unit list, but still a unit.)

**Reply by avee87** [reply]
Well, we already have this list here: https://github.com/home-assistant/core/blob/dev/homeassistant/util/unit_system.py#L86-L92
And it will probably never have things like voltage or power since pretty much everyone uses V and W for them. And even if we do decide to add them to this list, what's the harm in also having optional config fields for them?

**Reply by frenck** [reply]
> since pretty much everyone uses V and W for them

Not really. Most batteries in the small IoT device use mV (as an example).

Anyways, my above/initial comment holds at this point.

**Comment by gwww** [comment]
Coming from Canada, where unit system is totally wacky, I'm understanding of the point that @frenck makes about "speed of a car" vs "wind speed". In Canada we most often talk about food weights in imperial although we buy them in metric. People weight is imperial, except when dealing with doctors and the government. We most often refer to height (of a person) in imperial but talk about distance to another city in metric. That is but a couple of examples of our screwed up transition to metric that started in 1970!

I agree it is a big game to do this "right", especially if you consider just the couple of examples that I stated. I'd vote no, not now on this one and revisit down the road.

**Comment by emontnemery** [comment]
The way forward here is to allow the user to customize the unit of individual sensors.

This solves the problem where there are multiple applications using different units without the need to add device classes for each of those applications. For example:
- The preferred unit of a distance may be mm / in for rainfall and km / mi for travel related distances
- The preferred unit of a pressure may be hPa / inHg for atmospheric pressure and bar / psi for tire pressure

This has so far been implemented for temperature and pressure sensors as well as for weather entities.

**Comment by knthm** [comment]
> I rather see the possibility to be able to change the unit on devices/entities.

I think this is the most scalable solution. Adding on to that idea: a global classification of unit types enforcing integration support for all units of a type to simplify integration usage of units in DRYable manner, as each component wouldn't need manual logic for interviewing user unit preference and then doing conversion via `util/FOO`.

In line with that:

> The preferred unit of a distance may be mm / in for rainfall

Some integrations currently don't support all types of a unit anyway, see following examples for rainfall:

- [components/weather](https://github.com/home-assistant/core/blob/dev/homeassistant/components/weather/__init__.py#L797-L817) currently converts between the two precipitation distance/hour units based on user settings.
- [components/aemet](https://github.com/home-assistant/core/blob/dev/homeassistant/components/aemet/const.py#L265-L269) only supports mm/hr.

Centralizing the enforcement of unit types for `PRECIPITATION_*_PER_HOUR` and having central logic for handling entity-provided unit preference with internal conversion would simplify the `weather` integration significantly while also enhancing the units returned by `aemet`.

---
To summarize the functionality:

Integrations would specify a preferred default unit class type for each entity, e.g. `PRECIPITATION_MILLIMETERS_PER_HOUR`.
This value would be user-overridable via dropdown in the entity settings dialog.

The value, device class, and user-provided unit type would be sent to a central handler for the unit class.
The handler would decide upon the converted value + unit to be returned:
- User-defined unit is final overriding all
- No user-defined unit -> default setting is modulated by global Imperial/Metric unit setting

Device classes should be tied to unit classes instead of defined by-hand as it currently stands. Device class units don't seem to be enforced by any code/tests anyway, as the only source restricting units per device class seems to be the [developer documentation](https://developers.home-assistant.io/docs/core/entity/sensor/#available-device-classes), though please correct me if I'm wrong.
